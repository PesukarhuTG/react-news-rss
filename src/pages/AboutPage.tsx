import React from 'react';
import styled from 'styled-components';
import { Layout } from '../components';

const AboutPage: React.FC = () => {
  return (
    <Layout>
      <Wrapper>
        <Title>About us</Title>
        <Description>
          <b>Racoon news</b>: it is the latest breaking news and information on the top stories,
          weather, business, entertainment, politics, and more.
        </Description>
        <PrimaryText>You do not miss anything!</PrimaryText>
      </Wrapper>
    </Layout>
  );
};

const Wrapper = styled.div`
  margin: 0 auto;
`;

const Title = styled.h2`
  text-align: center;
  font-size: 30px;
  font-weight: 700;
`;

const Description = styled.p`
  margin-top: 30px;
  text-align: center;
  font-size: 18px;
`;

const PrimaryText = styled.p`
  margin-top: 30px;
  text-align: center;
  font-size: 22px;
  font-weight: 700;
`;

export default AboutPage;
