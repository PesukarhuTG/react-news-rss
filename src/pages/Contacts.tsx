import React from 'react';
import styled from 'styled-components';
import FormProps from '../types/Form';
import { Layout, LoginForm, FormCardsAlbum } from '../components';

interface State {
  list?: FormProps[];
  message?: string;
}

class Contacts extends React.Component<State> {
  state = {
    list: [],
    message: '',
  };

  onSubmit = (formFields: FormProps) => {
    const currentList = this.state.list as FormProps[];
    currentList.push(formFields);
    this.setState({ list: currentList });

    this.setState({ message: 'Your data has been saved!' });
    setTimeout(() => this.setState({ message: '' }), 1500);
  };

  render() {
    return (
      <Layout>
        <Wrapper>
          <Title>Contact us via form</Title>
          <LoginForm onSubmit={this.onSubmit} />
        </Wrapper>
        <Message data-testid="form-message">{this.state.message}</Message>
        <FormCardsAlbum list={this.state.list} />
      </Layout>
    );
  }
}

const Wrapper = styled.div`
  margin: 0 auto;
`;

const Title = styled.h2`
  text-align: center;
  font-size: 30px;
  font-weight: 700;
`;

const Message = styled.p`
  height: 30px;
  font-size: 14px;
  font-weight: 700;
  text-align: center;
  color: var(--accent);
`;

export default Contacts;
