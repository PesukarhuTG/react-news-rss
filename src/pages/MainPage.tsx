import React, { useState, useEffect } from 'react';
import {
  Layout,
  SearchPanel,
  CardsAlbum,
  SortSelectBy,
  SortSelectIn,
  SortDateFrom,
  SortDateTo,
  Pagination,
} from 'components';
import styled from 'styled-components';
import { getNews, searchNews } from '../services/getDataApi';
import CardProps from '../types/Card';
import useNewsContext from '../store/Context';

const MainPage: React.FC = () => {
  const [news, setNews] = useState<CardProps[]>([]);
  const [message, setMessage] = useState<string>('');

  const {
    searchVal,
    searchIn,
    sortBy,
    sortDateFrom,
    sortDateTo,
    currentPage,
    setCurrentPage,
    totalPageAmount,
    setTotalPageAmount,
    pageSize,
    setPageSize,
  } = useNewsContext();

  const onSubmit = (data: CardProps[]): void => {
    if (!data.length) {
      setMessage('Sorry, your request is failed');
    } else {
      setMessage('');
    }
    setNews(data);
  };

  useEffect(() => {
    const checkData = async (): Promise<void> => {
      if (searchVal) {
        const data = await searchNews(
          searchVal,
          searchIn,
          sortBy,
          sortDateFrom,
          sortDateTo,
          currentPage,
          pageSize
        ).then((resp) => resp);

        if (!data.articles.length) {
          setMessage('Sorry, your request is failed');
        } else {
          setMessage('');
          setNews(data.articles);
          setTotalPageAmount(data.totalResults > 100 ? 100 : data.totalResults);
        }
      } else {
        const data = await getNews(currentPage, pageSize).then((resp) => resp);
        setNews(data.articles);
        setTotalPageAmount(data.totalResults > 100 ? 100 : data.totalResults);
      }
    };

    checkData();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchIn, sortBy, sortDateFrom, sortDateTo, currentPage, pageSize]);

  return (
    <Layout>
      <SearchPanel onSearch={onSubmit} />
      <Headling>Hot news on Racoon digest</Headling>
      <SortWrapper>
        <SortBlock>
          <p>Searching by:</p>
          <SortSelectIn options={['title', 'description']} />
          <SortSelectBy options={['publishedAt', 'relevancy', 'popularity']} />
        </SortBlock>
        <SortBlock>
          <SortDateFrom />
          <SortDateTo />
        </SortBlock>
      </SortWrapper>
      <Message data-testid="fail-message">{message}</Message>
      <CardsAlbum cards={news} />
      <Pagination
        page={currentPage}
        onChange={(page, pageSize) => {
          setCurrentPage(page);
          setPageSize(pageSize);
        }}
        total={totalPageAmount}
        pageSize={pageSize}
      />
    </Layout>
  );
};

const Headling = styled.p`
  margin: 20px 0 10px;
  text-align: center;
  font-size: 36px;
  font-weight: 700;
  text-transform: uppercase;
  color: var(--primary);
  opacity: 0.6;
`;

const Message = styled.p`
  margin: 10px 0 10px;
  height: 20px;
  text-align: center;
  font-size: 16px;
  font-weight: 700;
  color: var(--accent);
`;

const SortBlock = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 20px;
`;

const SortWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 20px;
  gap: 20px;
  background-color: var(--second-contrast);
  font-size: 14px;

  @media (max-width: 950px) {
    flex-direction: column;
  }
`;

export default MainPage;
