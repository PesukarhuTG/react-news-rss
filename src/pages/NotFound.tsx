import React from 'react';
import styled from 'styled-components';
import { Layout } from '../components';
import { Link } from 'react-router-dom';

const NotFound: React.FC = () => {
  return (
    <Layout>
      <Wrapper>
        <Title>404</Title>
        <Description>sorry, page not found</Description>
        <HomeLink to="/">â–º back home</HomeLink>
      </Wrapper>
    </Layout>
  );
};

const Wrapper = styled.div`
  margin: auto;
`;

const Title = styled.h2`
  text-align: center;
  font-size: 140px;
  font-weight: 700;
`;

const Description = styled.p`
  text-align: center;
  font-size: 24px;
  letter-spacing: 1.5px;
  margin: 0 0 20px;
`;

const HomeLink = styled(Link)`
  display: block;
  text-align: center;
  color: var(--primary);

  &:hover {
    text-decoration: underline;
  }
`;

export default NotFound;
