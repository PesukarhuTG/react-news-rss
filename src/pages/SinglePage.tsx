import React, { useEffect } from 'react';
import styled from 'styled-components';
import { Layout } from '../components';
import { useNavigate } from 'react-router-dom';
import useNewsContext from '../store/Context';
import SavedCardProps from '../types/SavedCardData';

const SinglePage: React.FC = () => {
  const navigate = useNavigate();
  const { savedCardData, setDisableCurrentPosition } = useNewsContext();

  const { author, description, publishedAt, title, urlToImage, url } =
    savedCardData as SavedCardProps;

  const backMainPage = () => {
    navigate('/', { replace: true });
  };

  useEffect(() => {
    if (!author && !description && !publishedAt && !title && !urlToImage && !url) {
      navigate('/', { replace: true });
    }

    return () => {
      setDisableCurrentPosition(true);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <Layout>
      <Wrapper>
        <NewsImage
          style={{
            backgroundImage: `url(${urlToImage}`,
          }}
        />
        <NewsFullTitle>{title}</NewsFullTitle>

        <NewsFullDescription>
          {description || 'Sorry, this news has no description'}
        </NewsFullDescription>
        <InfoWrapper>
          <PublishedDiv>
            <span>
              <b>Published date</b>
            </span>
            <span>{publishedAt ? publishedAt.slice(0, 10) : 'no date'}</span>
          </PublishedDiv>
          <PublishedDiv>
            <span>
              <b>Author</b>
            </span>
            <span>{author || 'incognita'}</span>
          </PublishedDiv>
        </InfoWrapper>
        <InfoWrapper>
          <Button onClick={backMainPage}>❮ Back to main page</Button>
          <a href={url} target="_blank" rel="noreferrer noopener">
            <Button>Link to full news ❯</Button>
          </a>
        </InfoWrapper>
      </Wrapper>
    </Layout>
  );
};

const Wrapper = styled.div`
  max-width: 500px;
  width: 100%;
  margin: 0 auto;
`;

const InfoWrapper = styled.div`
  width: 100%;
  display: flex;
  font-size: 14px;
  gap: 20px;
  margin: 20px 0;
  align-items: center;
  justify-content: space-between;
`;

const NewsImage = styled.div`
  width: 100%;
  height: 240px;
  margin: 40px 0 30px;
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;
  border-radius: 10px;
  box-shadow: 0 0 15px 0 var(--primary);
`;

const NewsFullTitle = styled.p`
  margin-bottom: 20px;
  text-align: center;
  font-size: 18px;
  font-weight: 700;
  color: var(--primary);
  text-transform: uppercase;
`;

const NewsFullDescription = styled.p`
  font-size: 14px;
  padding-bottom: 20px;
  border-bottom: 1px dotted var(--main-color);
`;

const Button = styled.button`
  display: block;
  max-width: 170px;
  width: 100%;
  padding: 10px;
  font-size: 14px;
  border-radius: 5px;
  background-color: var(--primary);
  border: 0;
  cursor: pointer;
  transition: 0.3s;
  &:hover {
    background-color: var(--button-hover);
  }
`;

const PublishedDiv = styled.div`
  display: flex;
  flex-direction: column;
  gap: 10px;
`;

export default SinglePage;
