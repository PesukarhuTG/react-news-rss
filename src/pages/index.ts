export { default as MainPage } from './MainPage';
export { default as AboutPage } from './AboutPage';
export { default as NotFound } from './NotFound';
export { default as FormPage } from './FormPage';
export { default as SinglePage } from './SinglePage';
