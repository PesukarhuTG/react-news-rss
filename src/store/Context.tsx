import React, { createContext, useReducer, useContext } from 'react';
import FormProps from '../types/Form';
import SavedCardProps from '../types/SavedCardData';
import InitialStateProps from '../types/InitialState';
import { initialState, Reducer } from '../store/Reducer';

interface Props {
  children?: React.ReactNode;
}

const Context = createContext<InitialStateProps>(initialState);

export const Provider = ({ children }: Props) => {
  const [state, dispatch] = useReducer(Reducer, initialState);

  const setSearchValue = (searchVal: string) => {
    dispatch({
      type: 'CHANGE_SEARCH_VALUE',
      payload: {
        searchVal: searchVal,
      },
    });
  };

  const setFormName = (formName: string) => {
    dispatch({
      type: 'CHANGE_FORM_NAME',
      payload: {
        formName: formName,
      },
    });
  };

  const setFormDate = (formDate: string) => {
    dispatch({
      type: 'CHANGE_FORM_DATE',
      payload: {
        formDate: formDate,
      },
    });
  };

  const setFormCity = (formCity: string) => {
    dispatch({
      type: 'CHANGE_FORM_CITY',
      payload: {
        formCity: formCity,
      },
    });
  };

  const setFormAccept = (formAccept: boolean) => {
    dispatch({
      type: 'CHANGE_FORM_ACCEPT',
      payload: {
        formAccept: formAccept,
      },
    });
  };

  const setFormGender = (formGender: string) => {
    dispatch({
      type: 'CHANGE_FORM_GENDER',
      payload: {
        formGender: formGender,
      },
    });
  };

  const setFormList = (formList: FormProps[]) => {
    dispatch({
      type: 'CHANGE_FORM_LIST',
      payload: {
        formList: formList,
      },
    });
  };

  const setSearchIn = (searchIn: string) => {
    dispatch({
      type: 'CHANGE_SEARCH_IN',
      payload: {
        searchIn: searchIn,
      },
    });
  };

  const setSortBy = (sortBy: string) => {
    dispatch({
      type: 'CHANGE_SORT_BY',
      payload: {
        sortBy: sortBy,
      },
    });
  };

  const setSortDateFrom = (sortDateFrom: string) => {
    dispatch({
      type: 'CHANGE_SORT_DATE_FROM',
      payload: {
        sortDateFrom: sortDateFrom,
      },
    });
  };

  const setSortDateTo = (sortDateTo: string) => {
    dispatch({
      type: 'CHANGE_SORT_DATE_TO',
      payload: {
        sortDateTo: sortDateTo,
      },
    });
  };

  const setCurrentPage = (currentPage: number) => {
    dispatch({
      type: 'CHANGE_CURRENT_PAGE',
      payload: {
        currentPage: currentPage,
      },
    });
  };

  const setTotalPageAmount = (totalPageAmount: number) => {
    dispatch({
      type: 'CHANGE_TOTAL_PAGE_AMOUNT',
      payload: {
        totalPageAmount: totalPageAmount,
      },
    });
  };

  const setPageSize = (pageSize: number) => {
    dispatch({
      type: 'CHANGE_PAGE_SIZE',
      payload: {
        pageSize: pageSize,
      },
    });
  };

  const setSelectedFile = (selectedFile: null | string) => {
    dispatch({
      type: 'CHANGE_FILE',
      payload: {
        selectedFile: selectedFile,
      },
    });
  };

  const setSavedCardData = (savedCardData: SavedCardProps) => {
    dispatch({
      type: 'CHANGE_SAVED_CARD_DATA',
      payload: {
        savedCardData: savedCardData,
      },
    });
  };

  const setDisableCurrentPosition = (disableCurrentPosition: boolean) => {
    dispatch({
      type: 'CURRENT_POSITION_INFO',
      payload: {
        disableCurrentPosition: disableCurrentPosition,
      },
    });
  };

  const setFileText = (fileText: string) => {
    dispatch({
      type: 'CHANGE_FILE_TEXT',
      payload: {
        fileText: fileText,
      },
    });
  };

  const value: InitialStateProps = {
    searchVal: state.searchVal,
    formName: state.formName,
    formDate: state.formDate,
    formCity: state.formCity,
    formAccept: state.formAccept,
    formGender: state.formGender,
    formList: state.formList,
    searchIn: state.searchIn,
    sortBy: state.sortBy,
    sortDateFrom: state.sortDateFrom,
    sortDateTo: state.sortDateTo,
    currentPage: state.currentPage,
    totalPageAmount: state.totalPageAmount,
    pageSize: state.pageSize,
    selectedFile: state.selectedFile,
    fileText: state.fileText,
    savedCardData: state.savedCardData,
    disableCurrentPosition: state.disableCurrentPosition,
    setSearchValue,
    setFormName,
    setFormDate,
    setFormCity,
    setFormAccept,
    setFormGender,
    setFormList,
    setSearchIn,
    setSortBy,
    setSortDateFrom,
    setSortDateTo,
    setCurrentPage,
    setTotalPageAmount,
    setPageSize,
    setSelectedFile,
    setSavedCardData,
    setDisableCurrentPosition,
    setFileText,
  };

  return <Context.Provider value={value}>{children}</Context.Provider>;
};

const useNewsContext = () => {
  const context = useContext(Context);

  if (context === undefined) {
    throw new Error('useNewsContext must be used within Context');
  }

  return context;
};

export default useNewsContext;
