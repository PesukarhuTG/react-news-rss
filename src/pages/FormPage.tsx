import React, { useState } from 'react';
import styled from 'styled-components';
import FormProps from '../types/Form';
import { Layout, Form, FormCardsAlbum } from '../components';
import { useDispatch, useSelector } from 'react-redux';
import { changeFormList } from 'store/FormSlice';
import { AppDispatch, RootState } from 'store/Store';

const FormPage: React.FC = () => {
  const [message, setMessage] = useState<string>('');
  const { formList } = useSelector((state: RootState) => state.forms);
  const dispatch = useDispatch<AppDispatch>();

  const onSubmit = (formFields: FormProps) => {
    const currentList = [...formList];
    currentList.push(formFields);
    dispatch(changeFormList(currentList));

    setMessage('Your data has been saved!');
    setTimeout(() => setMessage(''), 1500);
  };

  return (
    <Layout>
      <Wrapper>
        <Title>Save your data</Title>
        <Form onSubmit={onSubmit} />
      </Wrapper>
      <Message data-testid="form-message">{message}</Message>
      <FormCardsAlbum list={formList} />
    </Layout>
  );
};

const Wrapper = styled.div`
  margin: 0 auto;
`;

const Title = styled.h2`
  text-align: center;
  font-size: 30px;
  font-weight: 700;
`;

const Message = styled.p`
  height: 30px;
  font-size: 14px;
  font-weight: 700;
  text-align: center;
  color: var(--accent);
`;

export default FormPage;
