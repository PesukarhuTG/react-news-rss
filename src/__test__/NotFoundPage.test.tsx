import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { BrowserRouter } from 'react-router-dom';
import { NotFound } from '../pages';
import { Provider } from 'react-redux';
import store from '../store/Store';

describe('NotFound page tests', () => {
  test('render NotFound page', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <NotFound />
        </BrowserRouter>
      </Provider>
    );

    const title = screen.getByText(/404/i);
    expect(title).toBeInTheDocument();
  });

  test('NotFound page contains link to Home', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <NotFound />
        </BrowserRouter>
      </Provider>
    );

    const homeLink = screen.getByText(/back home/i);
    expect(homeLink).toBeInTheDocument();
  });
});
