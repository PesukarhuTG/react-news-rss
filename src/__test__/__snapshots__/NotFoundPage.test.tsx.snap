// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotFound page tests NotFound page snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-dkzDqf dbxxpE"
        data-testid="layout-container"
      >
        <header
          class="sc-bczRLJ fDDTxG"
        >
          <a
            aria-current="page"
            class="logo active"
            href="/"
          >
            RACOON digest
          </a>
          <nav
            class="sc-gsnTZi fwfCFK"
          >
            <a
              aria-current="page"
              class="active-link"
              data-testid="mainpage-link"
              href="/"
            >
              Home
            </a>
            <a
              data-testid="aboutpage-link"
              href="/about"
            >
              About
            </a>
            <a
              data-testid="contactspage-link"
              href="/contacts"
            >
              Contacts
            </a>
          </nav>
        </header>
        <div
          class="sc-fLlhyt dyulgG"
        >
          <h2
            class="sc-bBrHrO jNTwPq"
          >
            404
          </h2>
          <p
            class="sc-ivTmOn jzgAIj"
          >
            sorry, page not found
          </p>
          <a
            class="sc-cxabCf bimnAM"
            href="/"
          >
            ► back home
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-dkzDqf dbxxpE"
      data-testid="layout-container"
    >
      <header
        class="sc-bczRLJ fDDTxG"
      >
        <a
          aria-current="page"
          class="logo active"
          href="/"
        >
          RACOON digest
        </a>
        <nav
          class="sc-gsnTZi fwfCFK"
        >
          <a
            aria-current="page"
            class="active-link"
            data-testid="mainpage-link"
            href="/"
          >
            Home
          </a>
          <a
            data-testid="aboutpage-link"
            href="/about"
          >
            About
          </a>
          <a
            data-testid="contactspage-link"
            href="/contacts"
          >
            Contacts
          </a>
        </nav>
      </header>
      <div
        class="sc-fLlhyt dyulgG"
      >
        <h2
          class="sc-bBrHrO jNTwPq"
        >
          404
        </h2>
        <p
          class="sc-ivTmOn jzgAIj"
        >
          sorry, page not found
        </p>
        <a
          class="sc-cxabCf bimnAM"
          href="/"
        >
          ► back home
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
